---
title: "Enhancing Chemotherapy Response Prediction via Matched Colorectal Tumor-Organoid Gene Expression Analysis and Network-Based Biomarker Selection"
subtitle: "TCGA-COAD"
author: "Wei Zhang, Hanchen Huang, Xi S.Chen"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    highlight: pygments
    theme: yeti
    toc: true
    number_sections: true
    df_print: paged
    code_download: true
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
# Initial settings
knitr::opts_chunk$set(echo = TRUE)

dir.base <- "../../../"
# Create and retrieve dir
GEO <- "COAD"
dir.data <- file.path(dir.base, "data/", GEO)
dir.data.raw <- file.path(dir.data, "raw")
dir.data.processed <- file.path(dir.data, "processed")
dir.data.clinical <- file.path(dir.data, "clinical")

for(p in grep("dir",ls(),value = T)) dir.create(get(p),recursive = TRUE,showWarnings = FALSE)
```

# Auxillary function

```{r message = F}
fn_ls <- list.files("../utility/", pattern = ".R", full.names = T)
for (f in fn_ls) source(f)
```

# Data preprocessing

## Download from GDC

```{r}
# Download from GDC
query <- GDCquery(
    project = "TCGA-COAD",
    data.category = "Transcriptome Profiling",
    data.type = "Gene Expression Quantification")

GDCdownload(query = query,
            directory = dir.data.raw)
COAD_exp <- GDCprepare(query = query,
                       directory = dir.data.raw,
                       save = T,
                       save.filename = file.path(dir.data.raw, "TCGA_COAD.rda"))
```

```{r}
# Clinical data
query <- query <- GDCquery(
    project = "TCGA-COAD", 
    data.category = "Clinical", 
    data.format = "bcr xml"
)
GDCdownload(query,
            directory = dir.data.raw)
```

```{r}
# Get clinical data
clinical <- GDCprepare_clinic(query, clinical.info = "patient", directory = dir.data.raw)

# Get drug information
clinical.drug <- GDCprepare_clinic(query, clinical.info = "drug", directory = dir.data.raw)
```


```{r eval = F}
# Save clinical data
readr::write_csv(
  clinical,
  file.path(dir.data.clinical, "TCGA_COAD_colorectal_cancer_clinical_data.csv")
)
```

```{r}
# Save drug type information
readr::write_csv(
  clinical.drug,
  file.path(dir.data.clinical, "TCGA_COAD_colorectal_cancer_drug_type_info.csv")
)
```

## Preprocessing

```{r}
# Load saved TCGA COAD data (Loaded name as data)
load(
  file.path(dir.data.raw, "TCGA_COAD.rda")
)
```

```{r eval = F}
# Get fpkm 
exp <- assays(data)$fpkm_uq_unstrand

# Select Gene symbol and select protein coding
ref <- rowData(data) %>% data.frame()
protein_coding <- ref %>% 
  filter(gene_type == "protein_coding")

# Match with ENSG
common_id <- intersect(protein_coding$gene_id,
                       rownames(exp))
protein_coding <- protein_coding[match(common_id, protein_coding$gene_id),]
exp <- exp[common_id,]
rownames(exp) <- protein_coding$gene_name

# Match patient ID and select tumor samples
pheno <- colData(data) %>% 
  data.frame() %>%
  dplyr::filter(sample_type == "Primary Tumor")

# Remove duplicated sample id
pheno <- pheno[!pheno$is_ffpe,]
pheno <- pheno %>% 
  group_by(patient) %>%
  mutate(duplicated = ifelse(n() >= 2, T, F), .before = 1)
# Select duplicate samples with later plate number
pheno_dup <- pheno %>%
  filter(duplicated) %>%
  group_by(patient) %>%
  filter(str_sub(barcode, 22,22) == "A")
pheno <- rbind(pheno %>% filter(!duplicated), pheno_dup)
```

```{r}
# Match with expression data
exp <- exp[,pheno$barcode]
colnames(exp) <- pheno$patient

# Save filtered expression data
save(
  exp,
  pheno,
  file = file.path(dir.data.processed, "TCGA_COAD_colorectal_cancer_exp_primary_tumor_rm_dup.RDS")
)
```

```{r}
# Remove all 0s
exp <- gene_filtered(exp, log2_trans = T)
```

```{r eval = F}
# Save filtered expression data
write_csv(
  exp %>% data.frame() %>% rownames_to_column("Gene"),
  file = file.path(dir.data.processed, "TCGA_COAD_CC_exp_data_removed_0s_log2_trans.csv")
)
```

## Create overall survival information

```{r message = F}
# Select colorectal cancer with drup response
# Read clinical data
clinical <- read_csv(
  file.path(dir.data.clinical, "TCGA_COAD_colorectal_cancer_clinical_data.csv")
)

os_info <- data.table::fread(
  file.path(dir.data.raw, "tcga-COAD_survival.txt")
) %>% 
  mutate(bcr_patient_barcode =`_PATIENT`,
         .keep = "unused")

# Read expression data
exp <- read_csv(
  file.path(dir.data.processed, "TCGA_COAD_CC_exp_data_removed_0s_log2_trans.csv")
) %>% column_to_rownames("Gene")
colnames(exp) <- gsub("[.]", "-", colnames(exp))
```

```{r}
# Create survival data
clinical <- clinical %>%
  mutate(os = ifelse(vital_status == "Dead", 1, 0),
         os_day = ifelse(is.na(days_to_last_followup), days_to_death, days_to_last_followup)) %>%
  left_join(., os_info)
```

```{r}
# Match with expression samples
common_id <- intersect(clinical$bcr_patient_barcode, colnames(exp))
clinical <- clinical[match(common_id, clinical$bcr_patient_barcode),]
exp <- exp[,common_id]
```

```{r}
save(
  exp, 
  clinical,
  file = file.path(dir.data.processed, "TCGA_COAD_CC_matched_exp_clinical_data.rda")
)
```

# Session Information

```{r}
devtools::session_info()
```

